0000-crypto-construct-language.md

```diff
@@ -0,0 +1,761 @@
+document.
+
+For insance, in Secure Scuttlebutt, digital signatures are stored as:
+`"signature": "<base64>.sig.ed25519"`. The specification says to parse the
+JSON, remove the signature field, then canoncialize the JSON before hashing.
+With CCL, if the signature data starts at offest 152 and is 100 bytes long,
+then the CCL version of the signature would have two READ operations followed
+by a MERGE like so:
+
+```
+0 152 READ 251 $ READ MERGE
+```
+
+This would leave the stack with the JSON bytes before and after the signature
+line on the stack ready to be parsed, canonicalized, and then hashed. If the
+JSON is assumed to already be canoncicalized, then the hash can be directly
+calculated.
+
+### Multi-Sig Signatures
+
+Where it becomes obvious that CCL is superior to both Verifiable Credentials
+and Secure Scuttlebutt is in the encoding of multi-sig signatures. Multi-sig
+signatures are digital signatures constructed as a conglomeration of signatures
+from two or more identities. There are two kinds of multi-sig signatures:
+parallel and serial.
+
+Parallel signatures are when multiple identities sign the same data. This is a
+simple agreement protocol. Think of it like multiple parties signing the same
+contract. They are all agreeing to the data they are signing. Serial signatures
+are when one identity signs some data and then a second identity signs both the
+data concatenated with the first signature. The next signature would sign the
+concatenation of the data and all previous signatures. This is an endorsement
+protocol whereby each subsequent signature endorses the data and the previous
+signatures. This is useful in supervisory roles such as project maintainers in
+open source projects. Developers submit signed commits and then the maintainers
+sign the commit and the developer's signature when merging it into the main
+branch of the code, thus endorsing the contribution.
+
+Currently, the Verifiable Credentials spec doesn't directly specify how either
+type of multi-sig is to be serialized. It is left up to the implementor as long
+as each individual signature follows the spec, it is a compliant signature.
+
+In Secure Scuttlebutt protocol specification, multi-sig signatures are not
+addressed at all. There is some discussion on supporting this but no consensus
+has emerged mostly because all obvious solutions are ugly hacks and so far the
+need is low.
+
+With CCL, both kids of multi-sig signatures are trivial. By using an
+`IF-ELSE-THEN` opcode trio it is easy to append digital signatures to form
+a multi-sig signature. For instance, let's start with a basic digital
+signature that looks like the following:
+
+```
+<sig hex> HEX DECODE <pub key hex> HEX DECODE <nonce hex> HEX DECODE ED25519 DECRYPT foo.txt OPEN 0 $ READ CLOSE SHA256 HASH =
+```
+
+Let's say another signer wants to add a parallel signature to the one above.
+They would calculate their own signature and append it using `IF-ELSE-THEN` to
```

Comment:
Maybe this is some pre-existing convention somewhere (Forth? Stack machine?), but the IF-ELSE-THEN construct is confusing. I would think a IF-ELSE-FI (shell script) or IF-ELSE-ENDIF (ccp) would be less confusing.

Reply:
It is Forth: https://www.forth.com/starting-forth/4-conditional-if-then-statements/

I'm not married to the convention and we can totally adopt something more straightforward.

Reply:
I changed it to IF-ELSE-FI. Thanks for the suggestion.